### PR TITLE
feat(mem_cache): Hybrid Memory Pool system

### LINKED ISSUE #946: [Feature] Hybrid recurrent state management for KDA layers
## Goals

1. KDA recurrent state (accumulation buffer) and conv state (causal convolution sliding window) persist correctly across decode steps.
2. KV cache and recurrent state slot allocation/release/reuse are coordinated for hybrid models.
3. Recurrent state pool can be donated into JIT and correctly recovered after TPU computation.
4. Recurrent state is preserved across chunks during chunked prefill.
5. All existing models behave identically after interface migration (no regression).
6. Recurrent state buffers are correctly sharded under TP=1/4.

## Non-Goals

1. KDA kernel / hybrid backend implementation (follow-up task; this issue only provides the pool interface).
2. Decoder loop integration, weight loading.
3. RadixCache compatibility (follow-up task).

## Acceptance Criteria

1. KDA recurrent state and conv state persist correctly across decode steps.
2. KV cache and recurrent state slot allocation/release/reuse are coordinated for hybrid models.
3. Recurrent state pool can be donated into JIT and correctly recovered after TPU computation.
4. Recurrent state is preserved across chunks during chunked prefill.
5. All existing models behave identically after interface migration (no regression).
6. Recurrent state buffers are correctly sharded under TP=1/4.
7. Unit tests cover alloc/free/clear, dual buffer independence, slot reuse, and sharding.

## Task Breakdown

1. Implement `RecurrentStatePool` with dual buffers (recurrent f32 + conv bf16) and independent slot allocator.
2. Implement `HybridReqToTokenPool` to coordinate KV and recurrent state slot lifecycle.
3. Implement `MemoryPools` pytree container with JIT donate and `replace_all` support.
4. Add `recurrent_pool_indices` to `ForwardBatch` pipeline (`ScheduleBatch → ModelWorkerBatch → ForwardBatch`).
5. Align `ReqToTokenPool.alloc` signature with sglang PyTorch for chunked prefill slot reuse.
6. Add `has_recurrent_state` branch in `init_memory_pool` with memory budget estimation.
7. Migrate all model files from `token_to_kv_pool` to `memory_pools` interface.
8. Register `MLATokenToKVPool` as pytree node and implement `replace_buffer`.
9. Unit tests for alloc/free/clear, dual buffer independence, slot reuse, and sharding.

### PR DESCRIPTION
**Pull request by JamesBrianD**
## Summary

- Extract pool init pipeline (`init_memory_pool`, `profile_max_num_token`, `_compute_cell_size`) from `model_runner.py` into `ModelRunnerKVCacheMixin`, aligned with upstream sglang mixin pattern
- Add `RecurrentStatePool` with DP sharding for hybrid linear-recurrent models (Kimi-K2 style)
- Add `HybridReqToTokenPool` coordinating KV slot + recurrent state slot allocation, with per-DP free lists
- Add `MemoryPools` pytree container wrapping multiple pools with `replace_all()` dispatch
- Add recurrent budget split (`handle_recurrent_cache`) aligned with upstream `handle_max_mamba_cache`
- Rename `replace_kv_buffer` → `replace_buffer` across `KVCache` hierarchy for consistency
- Change `model_runner` JIT donate from `token_to_kv_pool` to `memory_pools`
- Add `recurrent_pool_idx` / `_chunked_slot_pinned` to `Req`, `recurrent_indices` to `ScheduleReqsInfo` / `ModelWorkerBatch` / `ForwardBatch` for hybrid recurrent data flow
- Add `is_hybrid_recurrent` flag to `ScheduleBatch` (isinstance detection, aligned with `is_hybrid` SWA pattern)
- Add `state_to_kv_ratio` server arg for recurrent/KV HBM budget split

## Test plan

- [x] Verify non-hybrid models (Qwen, DeepSeek, Llama) are unaffected — `MemoryPools` wraps single KV pool transparently
- [ ] Verify `MockModelRunner` creates `memory_pools` correctly
- [ ] Verify `ForwardBatch` pytree round-trips with `recurrent_indices=None` (non-hybrid) and with array (hybrid)
- [ ] Integration test with hybrid recurrent model config (future PR)

## Accuracy verification — MiMo-V2-Flash on TPU v7x

Verified non-hybrid model is unaffected by the memory-pool refactor.

**Setup**
- Branch: `feat/hybrid-memory-pool` @ `10786c9`
- Hardware: TPU v7x single-host (`2x2x1`, 4 chips / 8 logical devices)
- Model: `MiMo-V2-Flash` (FP8, 256 experts, hybrid SWA + full attention)
- Backend: `--moe-backend fused --attention-backend fa`
- Parallelism: `--tp-size 8 --dp-size 2 --ep-size 8`
- Other: `--page-size 256 --context-length 262144 --chunked-prefill-size 4096 --mem-fraction-static 0.95 --swa-full-tokens-ratio 0.25`

**AIME 2025 (EvalScope 0.17.1, AveragePass@1)**

| Subset | Num | Score |
|---|---|---|
| AIME2025-I | 15 | 0.8667 |
| AIME2025-II | 15 | 1.0000 |
| **Overall** | 30 | **0.9334** |

Generation config: `temperature=1, top_p=0.95, max_tokens=131072, enable_thinking=True`.

### CODE REVIEW

File: python/sgl_jax/srt/configs/kimi_linear.py
Comment (cjx0709) [inline python/sgl_jax/srt/configs/kimi_linear.py:1]: is this file should be in this pr? Maybe hybrid memory pool can be merged without modeling.

File: python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+        else:
+            self.token_to_kv_pool = MHATokenToKVPool(
+                size=self.max_total_num_tokens,
+                page_size=self.page_size,
+                dtype=self.kv_cache_dtype,
+                head_num=self.model_config.get_total_num_kv_heads_with_replication(
+                    self.attention_tp_size
+                ),
+                head_dim=(self.model_config.head_dim + 127) // 128 * 128,
+                layer_num=self.model_config.num_hidden_layers,
```
Comment (cjx0709) [inline python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py]: should use real full attention layer nums?

File: python/sgl_jax/srt/mem_cache/recurrent_state_pool.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+            return
+
+        idx_arr = jnp.asarray(indices, dtype=jnp.int32)
+        with jax.set_mesh(self.mesh):
+            for layer in range(self.num_linear_recurrent_layers):
+                self.recurrent_buffers[layer] = self.recurrent_buffers[layer].at[idx_arr].set(0)
+                for inner in range(len(self.conv_buffers[layer])):
+                    self.conv_buffers[layer][inner] = (
+                        self.conv_buffers[layer][inner].at[idx_arr].set(0)
+                    )
```
Comment (aolemila) [inline python/sgl_jax/srt/mem_cache/recurrent_state_pool.py]: It looks like to be lack of dp consideration.

File: python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+        linear_recurrent_layer_ids=cfg.linear_layer_ids,
+        size=max_num_reqs,
+        num_heads=linear_attn_config["num_heads"],
+        head_dim=linear_attn_config["head_dim"],
+        conv_kernel_size=linear_attn_config["short_conv_kernel_size"],
+        mesh=mesh,
+        dp_size=dp_size,
+    )
+    hybrid_pool = HybridReqToTokenPool(
+        size=max_num_reqs + 1,
```
Comment (Rodrian7) [inline python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py]: size should be max_num_reqs (not +1) — the extra KV slot is unreachable since recurrent cap is N.

File: python/sgl_jax/srt/mem_cache/recurrent_state_pool.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        self.num_k_heads = num_k_heads
+        self.head_k_dim = head_k_dim
+        self.conv_kernel_size = conv_kernel_size
+
+        proj_v = num_heads * head_dim
+        proj_k = num_k_heads * head_k_dim
+        self.proj_size = proj_v + 2 * proj_k
+
+        # total_slots: size+1 (for dummy slot 0), ceil to dp_size
+        self.total_slots = _ceil_to(size + 1, dp_size)
```
Comment (aolemila) [inline python/sgl_jax/srt/mem_cache/recurrent_state_pool.py]: `self.total_slots = _ceil_to(size + dp_size - 1, dp_size)` ?

size=7, dp_size=4, 
`self.total_slots = _ceil_to(size + 1, dp_size)` -> 8, number of dummy slots is 4, number of valid slots is 4 which is fewer than required.

File: python/sgl_jax/srt/model_executor/model_runner.py [hunk 5]
```diff
@@ -792,7 +524,7 @@ def _set_kv_cache_after_forward(self, layers_kv_fused):
                 for layer_kv_fused in layers_kv_fused
             ]
 
-        self.token_to_kv_pool.replace_kv_buffer(layers_kv_fused)
+        self.memory_pools.replace_all(layers_kv_fused)
```
Comment (Rodrian7) [inline python/sgl_jax/srt/model_executor/model_runner.py]: In epic, `_set_kv_cache_after_forward` is deleted and replaced by an inline `self.memory_pools.replace_all(pool_updates)` in `_forward`,with `jitted_run_model` returning a `pool_updates` dict instead of the `layers_kv_fused` list. Any specific reason this PR keeps the old
function + list-shaped return instead of mirroring epic?

File: python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py (hunk 3 above)
Comment (aolemila) [inline python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py]: no +1 ?

File: python/sgl_jax/srt/model_executor/model_runner.py [hunk 6]
```diff
@@ -219,13 +213,13 @@ def jitted_run_model(
...
             forward_batch,
-            token_to_kv_pool,
+            memory_pools,
             logits_metadata,
         ):
             model_state = jax.tree_util.tree_unflatten(model_state_def, model_state_leaves)
             model = nnx.merge(model_def, model_state)
             with LoraBatchContext.set_batch(forward_batch):
-                return model(forward_batch, token_to_kv_pool, logits_metadata)
+                return model(forward_batch, memory_pools.token_to_kv_pool, logits_metadata)
```
Comment (Rodrian7) [inline python/sgl_jax/srt/model_executor/model_runner.py]: The wrapper, jit donate, and signature all pass `memory_pools` whole, but this line splits it. Any reason not to pass `memory_pools` like epic?

File: python/sgl_jax/srt/configs/kimi_linear.py
Comment (JamesBrianD) [reply python/sgl_jax/srt/configs/kimi_linear.py:1]: Deleted

File: python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py (hunk 1 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py]: Changed to use `self._kv_pool_layer_count()`

File: python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py (hunk 3 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py]: Fixed

File: python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+            ReqToTokenPool,
+            SWAKVPool,
+        )
+
+        has_recurrent_state = self.linear_recurrent_config is not None
+
+        # --- ReqToTokenPool ---
+        if self.req_to_token_pool is None:
+            if has_recurrent_state:
+                _, self.req_to_token_pool, _ = _build_hybrid_pools(
```
Comment (Rodrian7) [inline python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py]: Epic builds KV pool first, then calls this once with real `token_to_kv_pool` and uses `mp` directly. Here we pass `None` and discard `mp`, then rebuild MemoryPools later — both args/return end up dead. Any reason for the 3-phase split?

File: python/sgl_jax/srt/model_executor/model_runner.py (hunk 5 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/model_executor/model_runner.py]: `_set_kv_cache_after_forward` was removed and the logic is inlined in _forward.

File: python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py (hunk 3 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py]: Fixed

File: python/sgl_jax/srt/model_executor/model_runner.py (hunk 6 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/model_executor/model_runner.py]: All change to the epic style codes.

File: python/sgl_jax/srt/mem_cache/recurrent_state_pool.py (hunk 2 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/mem_cache/recurrent_state_pool.py]: As suggested in the chat, we will place the "set to zeros" logic in the runtime rather than the control plane. We will use the `has_initial_state=False` flag to mark the new state.

File: python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py (hunk 7 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/model_executor/model_runner_kv_cache_mixin.py]: Fixed. The 3-phase split was an artifact of early development — _build_hybrid_pools was called with token_to_kv_pool=None just to get req_to_token_pool, then MemoryPools was rebuilt manually after the KV pool was ready. Now aligned with epic: KV pool is built first, then a single  `_build_hybrid_pools(token_to_kv_pool=self.token_to_kv_pool)` call returns all three objects together.

File: python/sgl_jax/srt/mem_cache/recurrent_state_pool.py (hunk 4 above)
Comment (JamesBrianD) [reply python/sgl_jax/srt/mem_cache/recurrent_state_pool.py]: Fixed. Calculate the size in the profile memory pool and keep ensuring the validity of numerical values

**Review by aolemila — DISMISSED** [review]

**Review by Bob-Chen222 — COMMENTED** [review]
LGTM. Maybe just needs to pass all the CI tests

File: python/sgl_jax/srt/managers/schedule_batch.py [hunk 8]
```diff
@@ -1058,6 +1080,13 @@ def prepare_for_extend(self):
             info.extend_lens = extend_lens
             info.extend_input_logprob_token_ids = extend_input_logprob_token_ids
 
+            info.has_initial_state = np.array([r.is_chunked > 0 for r in reqs], dtype=np.bool_)
```
Comment (aolemila) [inline python/sgl_jax/srt/managers/schedule_batch.py:1083]: For chunked request, KDA needs the previous state. So the value needs to be False.

Comment (JamesBrianD) [reply python/sgl_jax/srt/managers/schedule_batch.py:1083]: `has_initial_state` default values are True, for prefill `is_chunked == 0` set value to False

**Review by aolemila — APPROVED** [review]
